turtle_tf2_cpp/src/turtle_tf2_broadcaster.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2021, Open Source Robotics Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <geometry_msgs/msg/transform_stamped.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+#include <tf2/LinearMath/Quaternion.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <turtlesim/msg/pose.hpp>
+
+#include <memory>
+#include <string>
+
+using std::placeholders::_1;
+
+class FramePublisher : public rclcpp::Node
+{
+public:
+  FramePublisher()
+  : Node("turtle_tf2_frame_publisher")
+  {
+    // Declare and acquire `turtlename` parameter
+    this->declare_parameter<std::string>("turtlename", "turtle");
+    this->get_parameter("turtlename", turtlename_);
+
+    // Initialize the transform broadcaster
+    tf_broadcaster_ =
+      std::make_unique<tf2_ros::TransformBroadcaster>(*this);
+
+    // Subscribe to a turtle{1}{2}/pose topic and call handle_turtle_pose
+    // callback function on each message
+    std::ostringstream stream;
+    stream << "/" << turtlename_.c_str() << "/pose";
+    std::string topic_name = stream.str();
+
+    subscription_ = this->create_subscription<turtlesim::msg::Pose>(
+      topic_name, 10,
+      std::bind(&FramePublisher::handle_turtle_pose, this, _1));
+  }
+
+private:
+  void handle_turtle_pose(const turtlesim::msg::Pose::SharedPtr msg) const
```

Comment:
I think, in general, we're moving away from using the `SharedPtr` member of messages. My understanding of this is that it is a little syntactic sugar that doesn't really add much value (almost the same length) and is not completely clear as to what is happening (it could do more than just be a shared pointer to the object, you have to look to double check).

I'd prefer using the following:
```suggestion
  void handle_turtle_pose(const std::shared_ptr<turtlesim::msg::Pose> msg) const
```

This would be true in any other case (remove using `SharedPtr`, `UniquePtr`, or whatever is being used and create it explicitly.